### PR TITLE
knockout.mapping better strong typing when mapping objects back and forth

### PR DESCRIPTION
**Pull request by ffMathy**
removed fromjs overload that no longer exists in latest knockout.mapping, and added much better type safety for "fromJS".

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not provide its own types, and you can not add them.
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, and `strictNullChecks` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [x] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [x] Delete the package's directory.
- [x] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by ffMathy** [comment]
Hmmmm! @borisyankov can you help here? I think TypeScript might be outdated on the build server or something for custom types to be supported.

It works fine in the TypeScript playground and locally in `.ts` files - perhaps it has something to do with the fact that it's `.d.ts` files?

**Comment by ffMathy** [comment]
@borisyankov just confirmed with `tsc` to compile just fine, and it works in separate `.d.ts` and `.ts` files using TypeScript 2.4.

**Comment by andy-ms** [comment]
Hi @ffMathy, you just need to add a `// TypeScript Version: 2.3` comment on line 5 of `index.d.ts`.

**Comment by andy-ms** [comment]
From travis:
```
knockout.mapping-tests.ts(50,1): error TS2554: Expected 1-3 arguments, but got 2.
knockout.mapping-tests.ts(57,20): error TS2554: Expected 1-3 arguments, but got 2.
```
That error message is silly, but I think it means that there are only 1 and 3 argument overloads but no 2 argument overload.

**Comment by RyanCavanaugh** [comment]
@ffMathy Please fix the failures indicated in the Travis CI log.

**Comment by ffMathy** [comment]
Now I get a new failure. Any idea of what the cause of this may be?
> knockout depends on knockout.mapping but has a lower required TypeScript version.

**Comment by ffMathy** [comment]
Alright, it's building. I think I may have figured it out.

**Comment by ffMathy** [comment]
Nope, I am lost again. What is the cause of this error? I have no idea.

File: types/knockout/index.d.ts [hunk 1]
```diff
@@ -1,10 +1,7 @@
 // Type definitions for Knockout v3.4.0
 // Project: http://knockoutjs.com
-// Definitions by: Boris Yankov <https://github.com/borisyankov/>
-//                 Igor Oleinikov <https://github.com/Igorbek/>
-//                 Clément Bourgeois <https://github.com/moonpyk/>
-//                 Matt Brooks <https://github.com/EnableSoftware>
-//                 Benjamin Eckardt <https://github.com/BenjaminEckardt>
+// Definitions by: Boris Yankov <https://github.com/borisyankov/>, Igor Oleinikov <https://github.com/Igorbek/>, Clément Bourgeois <https://github.com/moonpyk/>, Matt Brooks <https://github.com/EnableSoftware>, Benjamin Eckardt <https://github.com/BenjaminEckardt>, Mathias Lorenzen <https://github.com/ffMathy>
+// TypeScript Version: 2.3
```
Comment (andy-ms) [inline types/knockout/index.d.ts]: This should go *below* the next line.
Also, for lists this long the multi-line version is preferable.

Comment (ffMathy) [reply types/knockout/index.d.ts]: I don't understand this. Which line goes under which? I can't make it work. I tried numerous combinations.

Is this documented somewhere? Can you help me tell how it should look instead?

**Comment by ffMathy** [comment]
I don't understand how this can fail @andy-ms when I have followed your advice, and my package looks identical compared to similar packages that also have the "Typescript Version" prefix.

**Comment by ffMathy** [comment]
Finally! It's ready for merging!

**Comment by johnnyreilly** [comment]
Merged!

**Pull request merged**